src/crypto/generic-ops.h

```diff
@@ -33,14 +33,19 @@
 #include <cstddef>
 #include <cstring>
 #include <functional>
+#include <sodium/crypto_verify_32.h>
 
 #define CRYPTO_MAKE_COMPARABLE(type) \
 namespace crypto { \
   inline bool operator==(const type &_v1, const type &_v2) { \
-    return std::memcmp(&_v1, &_v2, sizeof(type)) == 0; \
+    if (sizeof(_v1) == 32) \
+      return crypto_verify_32((const unsigned char*)&_v1, (const unsigned char*)&_v2) == 0; \
+    if (sizeof(_v1) == 8) \
+      return (const uint64_t&)_v1 == (const uint64_t&)_v2; \
+    return !memcmp(&_v1, &_v2, sizeof(_v1)); \
```

Comment:
This is still not constant time for the case of size other than 32 or 8 bytes, right?

While this may not matter for current usage it is leaving a booby trap for some future usage.

Reply:
This is correct about 32/8 vs others. The operator doesn't make any claim to be constant time in the first place, so I don't see any booby trap here. If you want constant time, you wouldn't use an operator which does not claim to do so.

Reply:
I don't understand the purpose of this commit except to implement the operation in constant time. You seem to be suggesting either that the operator now implements (presumably because this is relied upon by its use cases) constant time but this is not 'claimed' or that the operator claims to be constant time but only for 8 and 32 bytes but not other sizes?

Either off these seems quite odd and brittle to me, although I guess the latter would be okay if clearly documented.

Reply:
That's a bit more complex, but duplicates the macros so there's one specifically for constant time.

Reply:
> It just happens to do so for 32 bytes because we want it for secret keys and secret keys are 32 bytes

This is exactly what I am saying about relying on brittle behavior.

If we want constant time for private keys then there should a specific defined mechanism for getting that.